### PR TITLE
Buyer

### PR DESCRIPTION
**Pull request by jesoch**
Some scenarios do not work

### CODE REVIEW

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 1]
```diff
@@ -43,35 +44,35 @@ public void setup() {
...
     }
 
-    @And("^I post a purchase to advertisement \"([^\"]*)\"$")
-    public void iPostAPurchaseToAdvertisement(long advertisementId) throws Throwable {
+    @And("^I post a purchase to advertisement \"([^\"]*)\" who \"([^\"]*)\"$")
+    public void iPostAPurchaseToAdvertisement(long advertisementId, String username) throws Throwable {
         String message = "{ \"advertisements\": [\"/advertisement/" + advertisementId + "\"] }";
 
-        result = mockMvc.perform(post("/purchases")
+        result = mockMvc.perform(post("/users/" + username + "/purchases/")
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: Using `/users/<username>/purchases/` will not make the purchase more secure, or even better tested. What is more, the fact that this route works is already handled by spring, so we'd rather do it the other way (through `/purchases`) on which we are expecting Spring server to write the logged in user on the Purchase for us.
Remember that we are simulating an user (client), we do not want it to trust the username we give it to him. That is why with this `mockMvc` call we are also sending the authentication header credentials, which the server will check prior doing any purchase validating.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 2]
```diff
@@ -43,35 +44,35 @@ public void setup() {
                 .build();
     }
 
-    @And("^I post a purchase to advertisement \"([^\"]*)\"$")
-    public void iPostAPurchaseToAdvertisement(long advertisementId) throws Throwable {
+    @And("^I post a purchase to advertisement \"([^\"]*)\" who \"([^\"]*)\"$")
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: There is no need to pass in an username. See next review comment.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 3]
```diff
@@ -43,35 +44,35 @@ public void setup() {
...
     }
 
-    @And("^I post a mass purchase to advertisements \"([^\"]*)\" and \"([^\"]*)\"$")
-    public void iPostAMassPurchaseToAdvertisementsAnd(String firstAdvertisementId, String secondAdvertisementId) throws Throwable {
+    @And("^I \"([^\"]*)\" post a mass purchase to advertisements \"([^\"]*)\" and \"([^\"]*)\"$")
+    public void iPostAMassPurchaseToAdvertisementsAnd(String username,String firstAdvertisementId, String secondAdvertisementId) throws Throwable {
         String message = "{ \"advertisements\": [\"/advertisement/" + firstAdvertisementId + "\", \"/advertisement/" + secondAdvertisementId + "\"] }";
 
-        result = mockMvc.perform(post("/purchases")
+        result = mockMvc.perform(post("/users/"+username+"/purchases")
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: Same `/purchases` thing applies here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 4]
```diff
@@ -43,35 +44,35 @@ public void setup() {
...
     }
 
-    @And("^I post a purchase to no advertisement$")
-    public void iPostAPurchaseToNoAdvertisement() throws Throwable {
+    @And("^I \"([^\"]*)\" post a purchase to no advertisement$")
+    public void iPostAPurchaseToNoAdvertisement(String username) throws Throwable {
         String message = "{ \"advertisements\": [] }";
 
-        result = mockMvc.perform(post("/purchases")
+        result = mockMvc.perform(post("users/"+ username+ "/purchases")
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: ...and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 5]
```diff
@@ -96,7 +97,8 @@ public void thereIsAnAdvertisementWithPurchase() throws Throwable {
 
     @Then("^There is a purchase with (\\d+) advertisements$")
     public void thereIsAPurchaseWithAdvertisements(int numberAdvertisements) throws Throwable {
-        String advertisementLocation = "/purchases/1/advertisements";
+
+        String advertisementLocation = "/purchases/{username}/1/advertisements";
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: This route does not exist, and the same `/purchases` thing applies here as well.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 6]
```diff
@@ -107,7 +109,7 @@ public void thereIsAPurchaseWithAdvertisements(int numberAdvertisements) throws
 
     @And("^There is a purchase with advertisement title \"([^\"]*)\"$")
     public void thereIsAPurchaseWithAdvertisementTitle(String advertisementTitle) throws Throwable {
-        String advertisementLocation = "/purchases/1/advertisements";
+        String advertisementLocation = "/purchases/{username}/1/advertisements";
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: ...and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 7]
```diff
@@ -117,13 +119,13 @@ public void thereIsAPurchaseWithAdvertisementTitle(String advertisementTitle) th
...
     @And("^There is a purchase with purchaser \"([^\"]*)\"$")
-    public void thereIsAPurchaseWithPurchaser(String purchaser) throws Throwable {
+    public void thereIsAPurchaseWithPurchaser(User purchaser) throws Throwable {
         thereIsA("$.purchaser", purchaser);
     }
 
     @And("^There is a purchase with date$")
     public void thereIsAPurchaseWithDate() throws Throwable {
-        String advertisementLocation = "/purchases/1";
+        String advertisementLocation = "/purchases/{username}/1";
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 8]
```diff
@@ -140,7 +142,7 @@ public void thereIsAPurchaseWithTotal(double total) throws Throwable {
 
     @And("^There are (\\d+) purchases$")
     public void thereArePurchases(int numPurchases) throws Throwable {
-        result = mockMvc.perform(get("/purchases")
+        result = mockMvc.perform(get("/purchases/{username}")
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 9]
```diff
@@ -151,7 +153,7 @@ public void thereArePurchases(int numPurchases) throws Throwable {
     public void iPutPurchaseWithAdvertisement(String purchaseId, String advertisementId) throws Throwable {
         String message = "{ \"advertisements\": [\"/advertisement/" + advertisementId + "\"] }";
 
-        result = mockMvc.perform(put("/purchases/" + purchaseId)
+        result = mockMvc.perform(put("/purchases/{username}/" + purchaseId)
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 10]
```diff
@@ -163,7 +165,7 @@ public void iPutPurchaseWithAdvertisement(String purchaseId, String advertisemen
     public void iPatchPurchaseWithAdvertisement(String purchaseId, String advertisementId) throws Throwable {
         String message = "{ \"advertisements\": [\"/advertisement/" + advertisementId + "\"] }";
 
-        result = mockMvc.perform(put("/purchases/" + purchaseId)
+        result = mockMvc.perform(put("/purchases/{username}" + purchaseId)
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 11]
```diff
@@ -173,15 +175,15 @@ public void iPatchPurchaseWithAdvertisement(String purchaseId, String advertisem
 
     @And("^I delete purchase \"([^\"]*)\"$")
     public void iDeletePurchase(String purchaseId) throws Throwable {
-        result = mockMvc.perform(delete("/purchases/" + purchaseId)
+        result = mockMvc.perform(delete("/purchases/{username}/" + purchaseId)
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java [hunk 12]
```diff
@@ -173,15 +175,15 @@ public void iPatchPurchaseWithAdvertisement(String purchaseId, String advertisem
...
                 .contentType(MediaType.APPLICATION_JSON)
                 .accept(MediaType.APPLICATION_JSON)
                 .with(authenticate()))
                 .andDo(print());
     }
 
     private <T> void thereIsA(String jsonPath, T expected) throws Throwable {
-        String location = "/purchases/1";
+        String location = "/purchases/{username}/1";
```
Comment (zurfyx) [inline src/test/java/cat/udl/eps/softarch/steps/PurchaseStepDefs.java]: and here.

File: src/test/resources/features/Purchase.feature [hunk 13]
```diff
@@ -8,7 +8,7 @@ Feature: Purchase
     Given I login as "user" with password "password"
     And I post an advertisement with title "Santa" and price "2016"
     And I login as "user2" with password "password"
-    And I post a purchase to advertisement "1"
+    And I post a purchase to advertisement "1" who "user"
```
Comment (zurfyx) [inline src/test/resources/features/Purchase.feature]: All `who "user"` occurrences in this file should be removed for the previous `/purchases` reason.

File: src/main/java/cat/udl/eps/softarch/handler/PurchaseEventHandler.java [hunk 14]
```diff
@@ -43,17 +49,18 @@
...
 
         // The current purchaser; buying through accepting an offer will have a different handler.
-        purchase.setPurchaser(loggedInAs);
+        purchase.setPurchaser(purchaser);
 
         // A product cannot be purchased by their owner.
         for (Advertisement advertisement: purchase.getAdvertisements()) {
             Advertisement dbAdvertisement = advertisementRepository.findOne(advertisement.getId());
-            Assert.isTrue(!loggedInAs.equals(dbAdvertisement.getOwner()), "You cannot buy your own product.");
+            Assert.isTrue(!purchaser.getName().equals(dbAdvertisement.getOwner()), "You cannot buy your own product.");
```
Comment (zurfyx) [inline src/main/java/cat/udl/eps/softarch/handler/PurchaseEventHandler.java]: You are comparing the name with an `User` (remember that you have just changed `.getOwner()` to return an `User`).
Either do:
`!purchaser.equals(dbAdvertisement.getOwner())`
or
`!purchaser.getName().equals(dbAdvertisement.getOwner().getName())`

File: src/main/java/cat/udl/eps/softarch/domain/User.java [hunk 15]
```diff
@@ -29,11 +30,23 @@
     private String email;
     private Date birthday;
     private String country;
+    @NotEmpty(message = "Is necessary a password")
```
Comment (zurfyx) [inline src/main/java/cat/udl/eps/softarch/domain/User.java]: Password is not mandatory for OAuth logins afaik. If you force them to have a password, they might not be able to log in.

**Review by zurfyx — CHANGES_REQUESTED** [review]

File: src/test/resources/features/Purchase.feature (hunk 13 above)
Comment (rogargon) [reply src/test/resources/features/Purchase.feature]: @jesoch, @zurfyx is right, revert Purchase.feature and PurchaseStepDefs.java to previous content. The user associated to the purchase is not determined in the API URL, it is based on the user currently logged in. In this scenarios the user "user"...

**Pull request merged**